### Nama : Muchammad Yuda Tri Ananda 
### NIM  : 24060124110142
### Kelas  : D



Untuk mempelajari lebih lanjut mengenai properti dari nilai standard deviation dan variance,  silahkan terlebih dahulu download dataset (“Latihan.csv”) dari link berikut ini:

- https://s.id/aDGce  

Dataset “Latihan” merupakan dataset yang berisi sampel jumlah jawaban benar untuk masing
masing tipe soal dan skornya pada suatu Latihan untuk suatu mata kuliah tertentu.



Aturan Penulisan Kode Program: 
- Gunakan Google Collaboratory dan library Pandas untuk mengerjakan lembar kerja 
berikut. 
- Simpan DataFrame hasil pembacaan file dataset “Latihan.csv” ke dalam variable “df”



In [34]:
# Import Library
import pandas as pd

# Download File "Latihan.csv"
LINK_DATASET = "https://github.com/myudak/belajarUndip/blob/367836e9ccf8f699c44b920cebfd1b662624693e/SEMESTER_2/Statistika/Latihan.csv?raw=true"

# Simpan DataFrame ke variable df
df = pd.read_csv(LINK_DATASET)

# Isi dataframe top 5 baris
df.head()


Tipe1  Tipe2  Tipe3  Tipe4
0      5      5      4      2
1      5      3      2      3
2      5      4      2      3
3      5      3      2      2
4      4      3      2      4

## <u>Properti #1:</u> Menambah atau mengurangi dengan suatu nilai konstanta untuk semua dataset. 

Untuk menemukan properti pertama, kita akan mencari nilai  standard deviation dan variance dari kumpulan data, lalu menambahkan (atau mengurangi) suatu nilai konstanta yang sama untuk setiap baris dalam data, dan menemukan nilai  standard deviation dan variance  yang baru.

In [35]:
KOLOM = 'Tipe4'
tipe4 = df[KOLOM]
tipe4.head() # Melihatkan top 5 dataframe kolom Tipe4

0    2
1    3
2    3
3    2
4    4
Name: Tipe4, dtype: int64

### Langkah 1: Hitung nilai standard deviation dan variance sebelum memodifikasi dataset untuk variabel **Tipe4** (bebas memilih) 

In [36]:
tipe4_std = tipe4.std()
tipe4_var = tipe4.var()

print(f"Dataset kolom {KOLOM} sebelum penambahan konstanta")
print('Standar Deviasi\t:', tipe4_std)
print('Variansi\t:', tipe4_var)

Dataset kolom Tipe4 sebelum penambahan konstanta
Standar Deviasi	: 1.869439807751294
Variansi	: 3.4948051948051946


### Langkah 2: Modifikasi satu buah variabel dari dataset terkait dengan menambahkan atau mengurangi suatu konstanta untuk setiap data.

In [37]:
KONSTANTA = 69

In [38]:
tipe4_mod = df[KOLOM].map(lambda x: x + KONSTANTA)
tipe4_mod.head()

0    71
1    72
2    72
3    71
4    73
Name: Tipe4, dtype: int64

### Langkah 3: Hitung nilai standard deviation dan variance dari modified dataset untuk melengkapi tabel pada Langkah 1 dan lakukan analisa terhadap hasil yang diperoleh sesuai pertanyaan berikut ini. 

In [39]:
tipe4_mod_std = tipe4_mod.std()
tipe4_mod_var = tipe4_mod.var()

print(f"Dataset kolom {KOLOM} setelah penambahan konstanta")
print("Standar Deviasi\t:", tipe4_mod_std)
print("Variansi\t:", tipe4_mod_var)

Dataset kolom Tipe4 setelah penambahan konstanta
Standar Deviasi	: 1.869439807751294
Variansi	: 3.4948051948051946


| **Statistical Descriptive** | **Python Code**            | **Original Data**         | **Modified Data**         |
|----------------------------|----------------------------|---------------------------|---------------------------|
| *s* (Standard Deviation):                   | `tipe4.std()`             | 1.869439807751294        | 1.869439807751294         |
| *s²* (Variance):                      | `tipe4.var()`           | 3.4948051948051946                       | 3.4948051948051946                      |


| Python Code: |       ```df['Tipe4'].map(lambda x: x + 69)```        |
|--------------|---------------|
| Deskripsi:   | Menambahkan sebuah konstanta sebesar **69** untuk setiap data dari variabel **Tipe4** |

| Bagaimana penjumlahan/pengurangan suatu konstanta pada semua data mengubah nilai standard deviation dan variance?       |
|-----------------------------|
| Menambahkan atau mengurangi suatu konstanta pada semua data tidak akan memengaruhi nilai varians maupun standar deviasi, karena kedua ukuran ini bergantung pada seberapa jauh data tersebar dari rata-ratanya, bukan pada posisi absolut datanya. Saat semua nilai dalam data digeser naik atau turun dengan konstanta tertentu, rata-rata juga ikut bergeser sebesar konstanta itu, tetapi selisih antara setiap data dan rata-ratanya tetap sama. Karena varians dan standar deviasi dihitung berdasarkan selisih ini (deviasi dari rata-rata), maka nilainya tidak berubah. Dengan kata lain, distribusi data tetap memiliki bentuk dan lebar yang sama, hanya lokasinya saja yang bergeser. |

## <u>Properti #2:</u> Mengalikan atau membagi dengan suatu nilai konstanta untuk semua dataset. 

Mirip dengan properti pertama, apa yang akan terjadi ketika kita mengalikan atau membagi semua data dengan suatu nilai konstanta.

In [40]:
KOLOM = 'Tipe3'
tipe3 = df[KOLOM]
tipe3.head() # Melihatkan top 5 dataframe kolom Tipe4

0    4
1    2
2    2
3    2
4    2
Name: Tipe3, dtype: int64

### Langkah 1: Hitung nilai standard deviation dan variance sebelum memodifikasi dataset untuk variabel Tipe3 (bebas memilih)

In [41]:
tipe3_std = tipe3.std()
tipe3_var = tipe3.var()

print(f"Dataset kolom {KOLOM} sebelum penambahan konstanta")
print('Standar Deviasi\t:', tipe3_std)
print('Variansi\t:', tipe3_var)

Dataset kolom Tipe3 sebelum penambahan konstanta
Standar Deviasi	: 3.9046675366064867
Variansi	: 15.24642857142857


### Langkah 2: Modifikasi satu buah variabel dari dataset terkait dengan mengalikan suatu konstanta untuk setiap data. 

In [42]:
KONSTANTA = 69

In [43]:
tipe3_mod = df[KOLOM].map(lambda x: x * KONSTANTA)
tipe3_mod.head()

0    276
1    138
2    138
3    138
4    138
Name: Tipe3, dtype: int64

### Langkah 3: Hitung nilai standard deviation dan variance dari modified dataset untuk melengkapi tabel pada Langkah 1 dan lakukan analisa terhadap hasil yang diperoleh sesuai pertanyaan berikut ini. 

In [44]:
tipe3_mod_std = tipe3_mod.std()
tipe3_mod_var = tipe3_mod.var()

print(f"Dataset kolom {KOLOM} setelah penambahan konstanta")
print("Standar Deviasi\t:", tipe3_mod_std)
print("Variansi\t:", tipe3_mod_var)

Dataset kolom Tipe3 setelah penambahan konstanta
Standar Deviasi	: 269.4220600258476
Variansi	: 72588.24642857142


| **Statistical Descriptive** | **Python Code**            | **Original Data**         | **Modified Data**         |
|----------------------------|----------------------------|---------------------------|---------------------------|
| *s* (Standard Deviation):                   | `tipe3.std()`             | 3.9046675366064867        | 269.4220600258476         |
| *s²* (Variance):                      | `tipe3.var()`           | 15.24642857142857                      | 72588.24642857142                      |


| Python Code: |       ```df['Tipe4'].map(lambda x: x * 69)```        |
|--------------|---------------|
| Deskripsi:   | Mengalikan sebuah konstanta sebesar **69** untuk setiap data dari variabel **Tipe4** |

| Bagaimana perkalian/pembagian suatu konstanta pada semua data mengubah nilai standard deviation dan variance?       |
|-----------------------------|
| Jika seluruh data dikalikan atau dibagi dengan suatu konstanta $k$, maka **variance akan berubah sebesar $k^2$ kali** dan **standard deviation akan berubah sebesar $k$ kali** artinya, operasi ini memperbesar atau memperkecil penyebaran data secara proporsional terhadap konstanta tersebut, tetapi **tidak mengubah bentuk distribusi relatif antar data**. |

## Properti #3: Sensitif terhadap Outliers 

Gunakan data yang sama, apa yang akan terjadi jika kita terdapat siswa baru super genius dimana untuk setiap tipe ujian selalu mendapatkan poin maksimal yakni 1000.  

In [45]:
KOLOM = 'Tipe3'
tipe3 = df[KOLOM]
tipe3.head() # Melihatkan top 5 dataframe kolom Tipe4

0    4
1    2
2    2
3    2
4    2
Name: Tipe3, dtype: int64

### Langkah 1: Hitung nilai standard deviation dan variance sebelum memodifikasi dataset untuk variabel Tipe3 (bebas memilih) 

In [46]:
tipe3_std = tipe3.std()
tipe3_var = tipe3.var()

print(f"Dataset kolom {KOLOM} sebelum penambahan konstanta")
print('Standar Deviasi\t:', tipe3_std)
print('Variansi\t:', tipe3_var)

Dataset kolom Tipe3 sebelum penambahan konstanta
Standar Deviasi	: 3.9046675366064867
Variansi	: 15.24642857142857


### Langkah 2: Modifikasi dataset dengan menambahkan satu baris data.

In [47]:
DATA_BARU = 1000
tipe3_mod = pd.concat([df[KOLOM], pd.Series([DATA_BARU])])

### Langkah 3: Hitung nilai standard deviation dan variance dari modified dataset untuk melengkapi tabel pada Langkah 1 dan lakukan analisa terhadap hasil yang diperoleh sesuai pertanyaan berikut ini. 

In [48]:
tipe3_mod_std = tipe3_mod.std()
tipe3_mod_var = tipe3_mod.var()

print(f"Dataset kolom {KOLOM} setelah penambahan konstanta")
print("Standar Deviasi\t:", tipe3_mod_std)
print("Variansi\t:", tipe3_mod_var)

Dataset kolom Tipe3 setelah penambahan konstanta
Standar Deviasi	: 131.802847471152
Variansi	: 17371.99060150376


| **Statistical Descriptive** | **Python Code**            | **Original Data**         | **Modified Data**         |
|----------------------------|----------------------------|---------------------------|---------------------------|
| *s* (Standard Deviation):                   | `tipe3.std()`             | 3.9046675366064867        |  131.802847471152         |
| *s²* (Variance):                      | `tipe3.var()`           | 15.24642857142857                      | 17371.99060150376                   |


| Python Code: |       ```pd.concat([df["Tipe3"], pd.Series([1000])])```        |
|--------------|---------------|
| Deskripsi:   | Menambahkan satu baris data dengan nilai 1000 untuk variabel Tipe3 |

| Bagaimana pengaruh penambahan satu baris data baru berupa nilai ekstrem mengubah nilai standard deviation dan variance?     |
|-----------------------------|
| Menambahkan satu nilai ekstrem seperti 1000 ke dalam suatu dataset menyebabkan peningkatan yang signifikan pada nilai variance dan standard deviation karena nilai tersebut memperbesar jarak rata-rata terhadap titik-titik data lain, sehingga penyebaran keseluruhan data menjadi lebih lebar dan tidak lagi mewakili distribusi awal secara seimbang. Dengan kata lain, satu nilai outlier dapat merusak konsistensi ukuran dispersi dan membuat interpretasi data menjadi bias.|